### ISSUE TITLE
Report deployment status in Admin APIs and enable safely deleting drained deployments

### ISSUE THREAD

**Issue by pcholakov** [restatedev/restate#1958]
Currently, the list+get deployment Admin APIs do not report active status for a deployment the way the CLI does. Further, deployments can only be force-deleted via the Admin delete API. This leads to a confusing operator experience with the potential pitfall of being able to accidentally remove an active deployment. Drained deployments are safe to delete, but they can neither be easily discovered (outside of the CLI) nor safely deleted (as the DELETE operation requires `force=true`).

1. Add a deployment status attribute with possible values `Active` | `Drained` to at least the Get Deployment Admin API, and ideally the List Deployments API too.
2. Allow deleting fully-drained deployments without requiring `force=true`.

Implementing both of these would also enable an easy way to automatically prune fully drained deployments in bulk. This would be particularly helpful with Lambda deployments where you can easily accumulate large numbers of inactive deployments on older versions of a function.

Related: https://github.com/restatedev/restate/issues/1377

**Comment by slinkydeveloper** [comment]
Considering the usage people do, I think we generally need two apis:

* List of drained deployments, as simple as that. A deployment is considered drained if it's not the latest for a given service, and it has no invocations to it. This is what the operator needs, what user CI jobs need.
* On each deployment, its status. I think this one is handy, but i would personally not fetch it by default all the times. Only, let's say, fetch it if the user asks so with `--extra` in CLI or something like that.

**Comment by slinkydeveloper** [comment]
The query 

```
WITH active_deployments AS (
    /* Latest deployment_id for each service */
    SELECT DISTINCT deployment_id as id
    FROM sys_service
    WHERE deployment_id IS NOT NULL
    UNION
    /* pinned deployments on in-flight invocations */
    SELECT DISTINCT pinned_deployment_id as id
    FROM sys_invocation_status
    WHERE pinned_deployment_id IS NOT NULL AND status != 'completed'
)
/* All deployments */
SELECT d.id as id FROM sys_deployment d
EXCEPT
/* Minus active ones */
SELECT id from active_deployments;
```

Is good enough to figure drained deployments.

One open question that comes to my mind is giving some sort of "order" to drained deployments. This because some users might not want to immediately delete a drained deployment, but keep it around if they eventually wanna rollback to it (that ofc assuming we introduce the api to do so!)

**Comment by tillrohrmann** [comment]
Given the expensiveness of this query, it would be great if we could cache the result. However, caching the result begs the question how stale can the information be in order to work reliably. We already have the TOCTOU problem right now as users might restart an invocation on an older deployment after a user checked whether it's drained or not but with caching the result this problem becomes more severe.

Maybe one strategy for cache invalidation could be the following: The pre-requisite for a deployment to drain is that there are other deployments with newer service revisions. Once this is the case all newly ingested invocations will land on those deployments. So the cache could be invalidated once the `Schema` version changes + after `x` duration. There is still the problem with restarting an invocation on an explicit deployment, though. But maybe that is acceptable in terms of fuzziness.

**Comment by tillrohrmann** [comment]
Once we get to a point where obtaining the deployment status is cheap (e.g. if the caching works well or once we have efficient aggregation support), it would be great to also include the status in the `sys_deployment` table for the sake of completeness.

**Comment by slinkydeveloper** [comment]
> So the cache could be invalidated once the Schema version changes + after x duration. There is still the problem with restarting an invocation on an explicit deployment, though. But maybe that is acceptable in terms of fuzziness.

Agree, that's where i was getting at with the point of ordering too. you might want to keep the prev version even if drained for example. Let me think what we can do with the existing information, and we get it from there.

**Comment by tillrohrmann** [comment]
Regarding the query, it would be great to use the `sys_vqueues` table to find deployments that have pinned invocations if vqueues are enabled. I guess the query could look like:

```
WITH active_deployments AS (
    /* Latest deployment for each service */
    SELECT DISTINCT deployment_id AS id
    FROM sys_service
    WHERE deployment_id IS NOT NULL

    UNION

    /* Pinned deployments used by active VQueue invocations */
    SELECT DISTINCT deployment AS id
    FROM sys_vqueues
    WHERE deployment IS NOT NULL
      AND stage IN ('inbox', 'running', 'suspended', 'paused')
)
```

This will have the benefit that we won't read any finished invocations :-)

**Comment by tillrohrmann** [comment]
To reliably use the above mentioned query, we need to solve https://github.com/restatedev/restate/issues/5165.

**Comment by tillrohrmann** [comment]
We decided to not implement this feature for the time being.
